xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link XdsClient} instance encapsulates all of the logic for communicating with the xDS
+ * server. It may create multiple RPC streams for a series of xDS protocols (e.g., LDS, RDS,
+ * VHDS, CDS and EDS) over a single channel. Watch-based interfaces are provided for each set
+ * of data needed by gRPC.
+ *
+ * <p>This class should only be instantiated by the xDS resolver but can be passed to load
+ * balancing policies.
+ */
+abstract class XdsClient {
+
+  /**
+   * Data class containing the results of performing a series of resource discovery RPCs via
+   * LDS/RDS/VHDS protocols. The results may include configurations for path/host rewriting,
+   * traffic mirroring, retry or hedging, default timeouts and load balancing policy that will
+   * be used to generate a service config.
+   */
+  // TODO(chengyuanzhang): content TBD, most information comes from VirtualHost proto.
+  static final class ConfigUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class CdsUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class EdsUpdate {
+
+  }
+
```

Comment:
the LB implementation may not need it but can we also have an LdsUpdate please? The SDS (security) implementation needs it on the server side.

Reply:
LDS/RDS/VHDS will be hidden under the hood, only the final result, which is `ConfigUpdate` will be passed out. This is what we discussed recently with c++/go. But that discussion does not include anything about SDS. If the LDS response needs to be exposed for SDS, we may need to discuss with @markdroth.

Reply:
Sure. The `XdsClient` can have additional functionality beyond what is needed for LB and I think that should not be a problem

Reply:
I am fine with adding a new watcher API for the server side.  However, the specific details of the xDS protocol are intended to be hidden behind this API, so the new watcher should not be phrased as "an LDS watcher"; it should instead be phrased as "a server-side config watcher".  The fact that the server-side config currently comes from LDS is an implementation detail of the `XdsClient`, not something the gRPC server code needs to know or care about.

Reply:
sounds good. @voidzcy I will take care of adding this new API after the XdsClient interface is in and stable.